### PR TITLE
Detect node sass and react version

### LINKED ISSUE #13953: Next.js should detect when both `sass` and `node-sass` are installed and warn
# Feature request

Next.js should detect when a project has both `sass` and `node-sass` installed in their project.
This should emit a helpful warning.

## Describe the solution you'd like

We should instruct the user that one of these dependencies needs to be removed, preferably, `node-sass` (leaving `sass`).

### LINKED ISSUE #13961: Next.js dev server should warn if React is not at least 16.10
Next.js requires React 16.10 or newer for Fast Refresh to work correctly.

On older versions, Next.js will fallback to a full page reload every time.

We should emit a strong warning in this case and tell the user their dev experience will be suboptimal.

### PR DESCRIPTION
**Pull request by todortotev**
Closes [13961](https://github.com/vercel/next.js/issues/13961)
Closes [13953](https://github.com/vercel/next.js/issues/13953)

@Timer  @timneutkens Will cc you when done.

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 

**test/integration/app-functional/test/index.test.js** 
- Document and App > should not have any missing key warnings

<details>
<summary>Expand output</summary>

  ● Document and App › should not have any missing key warnings

    FetchError: request to http://localhost:40677/ failed, reason: connect ECONNREFUSED 127.0.0.1:40677

      at ClientRequest.<anonymous> (../node_modules/node-fetch/lib/index.js:1455:11)


  ● Test suite failed to run

    TypeError: Cannot read property 'pid' of undefined

      263 | export async function killApp(instance) {
      264 |   await new Promise((resolve, reject) => {
    > 265 |     treeKill(instance.pid, (err) => {
          |                       ^
      266 |       if (err) {
      267 |         if (
      268 |           process.platform === 'win32' &&

      at lib/next-test-utils.js:265:23
      at killApp (lib/next-test-utils.js:264:9)
      at integration/app-functional/test/index.test.js:26:18

</details>

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| buildDuration | 11.1s | 11.1s | ⚠️ +15ms |
| nodeModulesSize | 64.9 MB | 64.9 MB | ⚠️ +1.9 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 1.998 | 1.888 | -0.11 |
| / avg req/sec | 1251.23 | 1324.5 | ⚠️ +73.27 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.357 | 1.211 | -0.15 |
| /error-in-render avg req/sec | 1842.79 | 2065.15 | ⚠️ +222.36 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.49 kB | 6.49 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..ec24.js gzip | 10.5 kB | 10.5 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.9 kB | 56.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.58 kB | 5.58 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.92 kB | 6.92 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 52.4 kB | 52.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.3 kB | 26.3 kB | ✓ |
| Overall change | 26.3 kB | 26.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| _app.js gzip | 1.26 kB | 1.26 kB | ✓ |
| _error.js gzip | 3.25 kB | 3.25 kB | ✓ |
| hooks.js gzip | 881 B | 881 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.05 kB | 2.05 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 8.22 kB | 8.22 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| _app.module.js gzip | 604 B | 604 B | ✓ |
| _error.module.js gzip | 2.04 kB | 2.04 kB | ✓ |
| hooks.module.js gzip | 383 B | 383 B | ✓ |
| index.module.js gzip | 223 B | 223 B | ✓ |
| link.module.js gzip | 1.52 kB | 1.52 kB | ✓ |
| routerDirect..dule.js gzip | 281 B | 281 B | ✓ |
| withRouter.m..dule.js gzip | 278 B | 278 B | ✓ |
| Overall change | 5.33 kB | 5.33 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| _buildManifest.js gzip | 171 B | 171 B | ✓ |
| _buildManife..dule.js gzip | 180 B | 180 B | ✓ |
| Overall change | 351 B | 351 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| index.html gzip | 901 B | 901 B | ✓ |
| link.html gzip | 906 B | 906 B | ✓ |
| withRouter.html gzip | 894 B | 894 B | ✓ |
| Overall change | 2.7 kB | 2.7 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| buildDuration | 13.1s | 11.7s | -1.3s |
| nodeModulesSize | 64.9 MB | 64.9 MB | ⚠️ +1.9 kB |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.49 kB | 6.49 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..ec24.js gzip | 10.5 kB | 10.5 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.9 kB | 56.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.58 kB | 5.58 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.92 kB | 6.92 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 52.4 kB | 52.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.3 kB | 26.3 kB | ✓ |
| Overall change | 26.3 kB | 26.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| _app.js gzip | 1.26 kB | 1.26 kB | ✓ |
| _error.js gzip | 3.25 kB | 3.25 kB | ✓ |
| hooks.js gzip | 881 B | 881 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.05 kB | 2.05 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 8.22 kB | 8.22 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| _app.module.js gzip | 604 B | 604 B | ✓ |
| _error.module.js gzip | 2.04 kB | 2.04 kB | ✓ |
| hooks.module.js gzip | 383 B | 383 B | ✓ |
| index.module.js gzip | 223 B | 223 B | ✓ |
| link.module.js gzip | 1.52 kB | 1.52 kB | ✓ |
| routerDirect..dule.js gzip | 281 B | 281 B | ✓ |
| withRouter.m..dule.js gzip | 278 B | 278 B | ✓ |
| Overall change | 5.33 kB | 5.33 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| _buildManifest.js gzip | 171 B | 171 B | ✓ |
| _buildManife..dule.js gzip | 180 B | 180 B | ✓ |
| Overall change | 351 B | 351 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| _error.js | 874 kB | 874 kB | ✓ |
| 404.html | 4.13 kB | 4.13 kB | ✓ |
| hooks.html | 3.71 kB | 3.71 kB | ✓ |
| index.js | 875 kB | 875 kB | ✓ |
| link.js | 913 kB | 913 kB | ✓ |
| routerDirect.js | 905 kB | 905 kB | ✓ |
| withRouter.js | 905 kB | 905 kB | ✓ |
| Overall change | 4.48 MB | 4.48 MB | ✓ |

</details>

</details>

**Comment by TodorTotev** [comment]
@Timer  I believe this is ready for review. If you have the time to give me some directions on how I could check with different dependencies @ integration tests I would love to create the required tests!

**Comment by ijjk** [comment]
## Failing test suites 

**test/integration/api-catch-all/test/index.test.js** 
- API routes > Server support > should return data when catch-all
- API routes > Server support > should 404 when catch-all with index and trailing slash
- API routes > Server support > should return data when catch-all with index and no trailing slash
- API routes > Serverless support > should return data when catch-all
- API routes > Serverless support > should 404 when catch-all with index and trailing slash
- API routes > Serverless support > should return data when catch-all with index and no trailing slash
- API routes > dev support > should return data when catch-all
- API routes > dev support > should 404 when catch-all with index and trailing slash
- API routes > dev support > should return data when catch-all with index and no trailing slash

<details>
<summary>Expand output</summary>

  ● API routes › dev support › should return data when catch-all

    FetchError: request to http://localhost:45729/api/users/1 failed, reason: connect ECONNREFUSED 127.0.0.1:45729

      at ClientRequest.<anonymous> (../node_modules/node-fetch/lib/index.js:1455:11)

  ● API routes › dev support › should 404 when catch-all with index and trailing slash

    FetchError: request to http://localhost:45729/api/users/ failed, reason: connect ECONNREFUSED 127.0.0.1:45729

      at ClientRequest.<anonymous> (../node_modules/node-fetch/lib/index.js:1455:11)

  ● API routes › dev support › should return data when catch-all with index and no trailing slash

    FetchError: request to http://localhost:45729/api/users failed, reason: connect ECONNREFUSED 127.0.0.1:45729

      at ClientRequest.<anonymous> (../node_modules/node-fetch/lib/index.js:1455:11)

  ● API routes › Server support › should return data when catch-all

    command failed with code 1

      133 |         code !== 0
      134 |       ) {
    > 135 |         return reject(new Error(`command failed with code ${code}`))
          |                       ^
      136 |       }
      137 | 
      138 |       resolve({

      at ChildProcess.<anonymous> (lib/next-test-utils.js:135:23)

  ● API routes › Server support › should 404 when catch-all with index and trailing slash

    command failed with code 1

      133 |         code !== 0
      134 |       ) {
    > 135 |         return reject(new Error(`command failed with code ${code}`))
          |                       ^
      136 |       }
      137 | 
      138 |       resolve({

      at ChildProcess.<anonymous> (lib/next-test-utils.js:135:23)

  ● API routes › Server support › should return data when catch-all with index and no trailing slash

    command failed with code 1

      133 |         code !== 0
      134 |       ) {
    > 135 |         return reject(new Error(`command failed with code ${code}`))
          |                       ^
      136 |       }
      137 | 
      138 |       resolve({

      at ChildProcess.<anonymous> (lib/next-test-utils.js:135:23)

  ● API routes › Serverless support › should return data when catch-all

    command failed with code 1

      133 |         code !== 0
      134 |       ) {
    > 135 |         return reject(new Error(`command failed with code ${code}`))
          |                       ^
      136 |       }
      137 | 
      138 |       resolve({

      at ChildProcess.<anonymous> (lib/next-test-utils.js:135:23)

  ● API routes › Serverless support › should 404 when catch-all with index and trailing slash

    command failed with code 1

      133 |         code !== 0
      134 |       ) {
    > 135 |         return reject(new Error(`command failed with code ${code}`))
          |                       ^
      136 |       }
      137 | 
      138 |       resolve({

      at ChildProcess.<anonymous> (lib/next-test-utils.js:135:23)

  ● API routes › Serverless support › should return data when catch-all with index and no trailing slash

    command failed with code 1

      133 |         code !== 0
      134 |       ) {
    > 135 |         return reject(new Error(`command failed with code ${code}`))
          |                       ^
      136 |       }
      137 | 
      138 |       resolve({

      at ChildProcess.<anonymous> (lib/next-test-utils.js:135:23)


  ● Test suite failed to run

    TypeError: Cannot read property 'pid' of undefined

      272 | export async function killApp(instance) {
      273 |   await new Promise((resolve, reject) => {
    > 274 |     treeKill(instance.pid, (err) => {
          |                       ^
      275 |       if (err) {
      276 |         if (
      277 |           process.platform === 'win32' &&

      at lib/next-test-utils.js:274:23
      at killApp (lib/next-test-utils.js:273:9)
      at integration/api-catch-all/test/index.test.js:52:20

  ● Test suite failed to run

    TypeError: Cannot read property 'pid' of undefined

      272 | export async function killApp(instance) {
      273 |   await new Promise((resolve, reject) => {
    > 274 |     treeKill(instance.pid, (err) => {
          |                       ^
      275 |       if (err) {
      276 |         if (
      277 |           process.platform === 'win32' &&

      at lib/next-test-utils.js:274:23
      at killApp (lib/next-test-utils.js:273:9)
      at integration/api-catch-all/test/index.test.js:63:20

  ● Test suite failed to run

    TypeError: Cannot read property 'pid' of undefined

      272 | export async function killApp(instance) {
      273 |   await new Promise((resolve, reject) => {
    > 274 |     treeKill(instance.pid, (err) => {
          |                       ^
      275 |       if (err) {
      276 |         if (
      277 |           process.platform === 'win32' &&

      at lib/next-test-utils.js:274:23
      at killApp (lib/next-test-utils.js:273:9)
      at integration/api-catch-all/test/index.test.js:79:13

</details>

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| buildDuration | 11.1s | 12.3s | ⚠️ +1.3s |
| nodeModulesSize | 64.9 MB | 64.9 MB | ⚠️ +1.91 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 1.915 | 2.108 | ⚠️ +0.19 |
| / avg req/sec | 1305.54 | 1186.12 | -119.42 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.226 | 1.414 | ⚠️ +0.19 |
| /error-in-render avg req/sec | 2039.93 | 1767.87 | -272.06 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.49 kB | 6.49 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..f367.js gzip | 10.9 kB | 10.9 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 57.2 kB | 57.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.58 kB | 5.58 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 7.27 kB | 7.27 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 52.7 kB | 52.7 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.3 kB | 26.3 kB | ✓ |
| Overall change | 26.3 kB | 26.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.05 kB | 2.05 kB | ✓ |
| hooks.js gzip | 881 B | 881 B | ✓ |
| _app.js gzip | 1.26 kB | 1.26 kB | ✓ |
| _error.js gzip | 3.25 kB | 3.25 kB | ✓ |
| Overall change | 8.22 kB | 8.22 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| index.module.js gzip | 223 B | 223 B | ✓ |
| routerDirect..dule.js gzip | 281 B | 281 B | ✓ |
| withRouter.m..dule.js gzip | 278 B | 278 B | ✓ |
| _error.module.js gzip | 2.04 kB | 2.04 kB | ✓ |
| hooks.module.js gzip | 383 B | 383 B | ✓ |
| link.module.js gzip | 1.52 kB | 1.52 kB | ✓ |
| _app.module.js gzip | 604 B | 604 B | ✓ |
| Overall change | 5.33 kB | 5.33 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| _buildManifest.js gzip | 270 B | 270 B | ✓ |
| _buildManife..dule.js gzip | 275 B | 275 B | ✓ |
| Overall change | 545 B | 545 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| index.html gzip | 955 B | 955 B | ✓ |
| link.html gzip | 961 B | 961 B | ✓ |
| withRouter.html gzip | 948 B | 948 B | ✓ |
| Overall change | 2.86 kB | 2.86 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| buildDuration | 12.5s | 13.1s | ⚠️ +525ms |
| nodeModulesSize | 64.9 MB | 64.9 MB | ⚠️ +1.91 kB |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.49 kB | 6.49 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..f367.js gzip | 10.9 kB | 10.9 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 57.2 kB | 57.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.58 kB | 5.58 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 7.27 kB | 7.27 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 52.7 kB | 52.7 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.3 kB | 26.3 kB | ✓ |
| Overall change | 26.3 kB | 26.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.05 kB | 2.05 kB | ✓ |
| hooks.js gzip | 881 B | 881 B | ✓ |
| _app.js gzip | 1.26 kB | 1.26 kB | ✓ |
| _error.js gzip | 3.25 kB | 3.25 kB | ✓ |
| Overall change | 8.22 kB | 8.22 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| index.module.js gzip | 223 B | 223 B | ✓ |
| routerDirect..dule.js gzip | 281 B | 281 B | ✓ |
| withRouter.m..dule.js gzip | 278 B | 278 B | ✓ |
| _error.module.js gzip | 2.04 kB | 2.04 kB | ✓ |
| hooks.module.js gzip | 383 B | 383 B | ✓ |
| link.module.js gzip | 1.52 kB | 1.52 kB | ✓ |
| _app.module.js gzip | 604 B | 604 B | ✓ |
| Overall change | 5.33 kB | 5.33 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| _buildManifest.js gzip | 270 B | 270 B | ✓ |
| _buildManife..dule.js gzip | 275 B | 275 B | ✓ |
| Overall change | 545 B | 545 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | vercel/next.js canary  | TodorTotev/next.js detect-node-sass-and-react-version | Change |
| - | - | - | - |
| _error.js | 875 kB | 875 kB | ✓ |
| 404.html | 4.17 kB | 4.17 kB | ✓ |
| hooks.html | 3.79 kB | 3.79 kB | ✓ |
| index.js | 875 kB | 875 kB | ✓ |
| link.js | 915 kB | 915 kB | ✓ |
| routerDirect.js | 907 kB | 907 kB | ✓ |
| withRouter.js | 907 kB | 907 kB | ✓ |
| Overall change | 4.49 MB | 4.49 MB | ✓ |

</details>

</details>

File: packages/next/build/webpack-config.ts [hunk 1]
```diff
@@ -263,6 +263,26 @@ export default async function getBaseWebpackConfig(
     resolvedBaseUrl = path.resolve(dir, jsConfig.compilerOptions.baseUrl)
   }
 
+  if (isServer) {
```
Comment (timneutkens) [inline packages/next/build/webpack-config.ts:266]: This code should be moved to `packages/next/cli/next-dev.ts`

File: packages/next/build/webpack-config.ts [hunk 2]
```diff
@@ -263,6 +263,26 @@ export default async function getBaseWebpackConfig(
...
+    const packageJsonPath = path.resolve(dir, './package.json')
+    const packageJsonFile = parseJsonFile(packageJsonPath)
```
Comment (jamesmosier) [inline packages/next/build/webpack-config.ts:268]: Checking the version of React in their package.json might not be comprehensive enough. They could have `^16.8.6` but if they did their initial installation recently then they might actually have `16.13.1` installed as their React version (due to the version using the `^` which will bump them to the next minor version released).

I think checking their installed dependencies might give you a bit more coverage:

```js
const packagePath = require.resolve('react/package.json')
const { version } = require(packagePath)
```

Comment (todortotev) [reply packages/next/build/webpack-config.ts:268]: Noted! Will change it!

**Pull request closed without merging**